File: VB/WinWebSolution.Module/CustomFieldMappingsSchedulerListViewControllerBase.vb
Imports DevExpress.ExpressApp
Imports DevExpress.XtraScheduler
Imports System.Collections.Generic
Imports DevExpress.ExpressApp.Scheduler

Namespace WinWebSolution.Module
    Public MustInherit Class CustomFieldMappingsSchedulerListViewControllerBase
        Inherits ViewController(Of ListView)

        Public Const SchedulerListEditorActiveKey As String = "SchedulerListEditor is used for the ListView"
        Public Sub New()
            TargetObjectType = GetType(MyEvent)
        End Sub
        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            Active(SchedulerListEditorActiveKey) = SchedulerListEditor IsNot Nothing
        End Sub
        Public ReadOnly Property SchedulerListEditor() As SchedulerListEditorBase
            Get
                Return TryCast(View.Editor, SchedulerListEditorBase)
            End Get
        End Property
        Protected Overrides Sub OnViewControlsCreated()
            MyBase.OnViewControlsCreated()
            AddCustomFieldMappings(GetCustomFieldMappings())
        End Sub
        Protected Overridable Function GetCustomFieldMappings() As List(Of MappingBase)
            Dim mappings As New List(Of MappingBase)()
            mappings.Add(New AppointmentCustomFieldMapping("ReferenceField", "CustomReferenceTypeField"))
            mappings.Add(New AppointmentCustomFieldMapping("SimpleField", "CustomSimpleTypeField"))
            Return mappings
        End Function
        Protected MustOverride Sub AddCustomFieldMappings(ByVal mappings As List(Of MappingBase))
    End Class
End Namespace

File: CS/WinWebSolution.Module/CustomFieldMappingsSchedulerListViewControllerBase.cs
using System;
using DevExpress.ExpressApp;
using DevExpress.XtraScheduler;
using System.Collections.Generic;
using DevExpress.ExpressApp.Scheduler;

namespace WinWebSolution.Module {
    public abstract class CustomFieldMappingsSchedulerListViewControllerBase : ViewController<ListView> {
        public const string SchedulerListEditorActiveKey = "SchedulerListEditor is used for the ListView";
        public CustomFieldMappingsSchedulerListViewControllerBase() {
            TargetObjectType = typeof(MyEvent);
        }
        protected override void OnViewChanging(View view) {
            base.OnViewChanging(view);
            ListView lv = view as ListView;
            if (lv != null)
                Active[SchedulerListEditorActiveKey] = GetSchedulerListEditor(lv) != null;
        }
        private static SchedulerListEditorBase GetSchedulerListEditor(ListView lv) {
            return lv.Editor as SchedulerListEditorBase;
        }
        public SchedulerListEditorBase SchedulerListEditor {
            get { return GetSchedulerListEditor(View); }
        }
        protected override void OnViewControlsCreated() {
            base.OnViewControlsCreated();
            AddCustomFieldMappings(GetCustomFieldMappings());
        }
        protected virtual List<MappingBase> GetCustomFieldMappings() {
            List<MappingBase> mappings = new List<MappingBase>();
            mappings.Add(new AppointmentCustomFieldMapping("CustomReferenceTypeField", "CustomReferenceTypeField!"));
            mappings.Add(new AppointmentCustomFieldMapping("CustomSimpleTypeField", "CustomSimpleTypeField"));
            return mappings;
        }
        protected abstract void AddCustomFieldMappings(List<MappingBase> mappings);
    }
}